### PR TITLE
Added variant for optional papi components 

### PR DESCRIPTION
**Pull request by G-Ragghianti**
New multi-value variant includes support for example, cuda, nvml, infiniband, infiniband_umad, powercap, rapl, and lmsensors.   Added dependency on cuda when installing with optional components cuda or nvml.

### CODE REVIEW

**Comment by G-Ragghianti** [comment]
Adam, Yes I think so. The multi-value variant echos the underlying papi config option "--components=cuda,nvml,etc" which can accept multiple values.

File: var/spack/repos/builtin/packages/papi/package.py [hunk 1]
```diff
@@ -47,7 +47,38 @@ class Papi(Package):
...
+                    "--with-cuda-dir=%s" % spec['cuda'].prefix]
+                configure(*configure_args)
+        with working_dir("src/components/infiniband_umad"):
+            if 'components=infiniband_umad' in spec:
+                configure()
+        with working_dir("src/components/lmsensors"):
+            if 'components=lmsensors' in spec:
+                configure_args = [
+                    "--with-sensors_incdir=/usr/include/sensors",
+                    "--with-sensors_libdir=/usr/lib64"]
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/papi/package.py]: Are these external dependencies? You may have to add packages for these in Spack as they may not exist for everyone.

**Comment by G-Ragghianti** [comment]
I know that nvml depends only on cuda and this dependency is required when either the cuda or nvml components are requested.  The infiniband_umad and lmsensors components do depend on the system to provide these.  The PAPI dev team intends these optional components to be added if possible but to fail quietly if they cannot be built.  I'm not sure if this jives with the way spack wants to manage dependencies though.  What do you suggest?

**Comment by adamjstewart** [comment]
I guess the question is:

1. Is it possible to install these sensors with Spack?
2. Is it possible that these sensors are installed on the system but in a different location (`/usr/lib` instead of `/usr/lib64`)?

Ideally, this package will "just work" for everyone who uses it. That's why Spack works so hard to package dependencies that may be on some computers but not others. The fact that your script specifies where these sensors can be found leads me to believe that they could be installed anywhere.

P.S. None of these things are deal breakers, we can still merge this package as is. Just trying to prevent future work by frustrated users who can't get the package working on their machines.

**Comment by G-Ragghianti** [comment]
The addition of the "lm-sensors" package should satisfy the dependencies of the newly included papi components. Please let me know if this is sufficient to merge the changes.  Thanks.

**Comment by G-Ragghianti** [comment]
FYI: the build failure appears to be a problem with travis and not with my modifications.

File: var/spack/repos/builtin/packages/papi/package.py [hunk 2]
```diff
@@ -47,7 +47,41 @@ class Papi(Package):
...
+            'infiniband_umad', 'powercap', 'rapl', 'lmsensors'),
+        multi=True,
+        description='Include optional components')
+    depends_on('cuda', when='components=cuda')
+    depends_on('cuda', when='components=nvml')
+    depends_on('lm-sensors', when='components=lmsensors')
+
     def install(self, spec, prefix):
+        with working_dir("src/components/cuda"):
+            if 'components=cuda' in spec:
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/papi/package.py]: I would put the if-statements outside of the with-statements. No need to change directories if the particular component isn't requested.

File: var/spack/repos/builtin/packages/papi/package.py [hunk 3]
```diff
@@ -47,7 +47,41 @@ class Papi(Package):
...
+                configure_args = [
+                    "--with-cuda-dir=%s" % spec['cuda'].prefix,
+                    "--with-cupti-dir=%s/extras/CUPTI" % spec['cuda'].prefix]
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/papi/package.py]: Is `--prefix` not needed here?

File: var/spack/repos/builtin/packages/papi/package.py [hunk 4]
```diff
@@ -57,6 +91,10 @@ def install(self, spec, prefix):
             # fail, so that PAPI does not get confused
             configure_args.append('MPICC=:')
 
+            configure_args.append('--with-components={0}'.format(
+                ' '.join(spec.variants['components'].value))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/papi/package.py]: Does this work even if the default `''` is used? You would effectively be passing `--with-components=` to the `configure` invocation.

**Comment by G-Ragghianti** [comment]
We are trying to get the spack-installed cuda working with our cuda papi component.  Right now, the problem is that the cuda package is not installing the real libcuda.so into $CUDADIR/lib64.  It is only installing a stub in $CUDADIR/lib64/stubs/.   How is it recommended to work around this?

File: var/spack/repos/builtin/packages/papi/package.py (hunk 2 above)
Comment (G-Ragghianti) [reply var/spack/repos/builtin/packages/papi/package.py]: Thanks for the recommendation.  I've modified this as suggested.

**Pull request closed without merging**